### ISSUE TITLE
Support setting external-ip twice, for IPv4 and IPv6

### ISSUE THREAD

**Issue by maximbaz** [coturn/coturn#682]
Right now we can either set `external-ip` once (if no RFC5780 functionality is required), or use the extended syntax to map global to local IP addresses.

In environments like Docker where local IP address changes all the time, using the extended syntax is quite cumbersome, so it is nice to just define `external-ip=<public-ip>` and it will map correctly every time.

However because I have two public IP addresses (one IPv4 and another IPv6), I would like to map them both using this syntax.

Could `coturn` accept not just one `external-ip`, but one of each type, so two in total, and map correspondingly public IPv4 addr to internal IPv4 address, and IPv6 to IPv6?

Following up on this question, given that I have two public IP addresses (albeit of different types), is it actually possible to make use of RFC5780 functionality? Or it requires two public IP addresses of exactly the same type? Lets assume we are talking about clients who support IPv6.

And finally a third question, given that it is now possible to use domain name in the `external-ip` field (as explained [here](https://github.com/coturn/coturn/issues/563#issuecomment-747190232)), could we maybe build on top of this, and map the first resolved A record to `external-ip` v4 and the first resolved AAAA record to `external-ip` v6? That way we can specify `external-ip=turn.example.com` and this would use both IPv4 and IPv6 if A and AAAA records are set for `turn.example.com` domain?

Thank you 🙂

**Comment by misi** [comment]
I am busy with the release but after it I will get back to you.

**Comment by misi** [comment]
> Right now we can either set `external-ip` once (if no RFC5780 functionality is required), or use the extended syntax to map global to local IP addresses.
> 
> In environments like Docker where local IP address changes all the time, using the extended syntax is quite cumbersome, so it is nice to just define `external-ip=<public-ip>` and it will map correctly every time.
AFAIK docker assign IP when it is starting a docker container. So e.g. in ENTYPOINT shell script you have in environment var  the local ip  by default, and maybe you can also could get as extra ENV var the global IP too, and can replace in config before starting coturn.
Also in k8s you can use multus where you can listen global IP directly.
> 
> However because I have two public IP addresses (one IPv4 and another IPv6), I would like to map them both using this syntax.
In IPv6 NAT is very uncommon and even early days of IPv6 there was no NAT AFAIR.
So IPv6 address as NAT-ed is very uncommon AFAIK.

```
For reasons discussed in [RFC2993] and Section 5, the IETF does not
recommend the use of Network Address Translation technology for IPv6.
```
If it is real need, and there is strong use case, then we can create a separated external ip for IPv4 and IPv6, but if possible don't NAT IPv6.
> 
> Could `coturn` accept not just one `external-ip`, but one of each type, so two in total, and map correspondingly public IPv4 addr to internal IPv4 address, and IPv6 to IPv6?
> 
IPv6 NAT is not very uncommon and also not recommended in RFC-s, but technically I guess we can do it.
> Following up on this question, given that I have two public IP addresses (albeit of different types), is it actually possible to make use of RFC5780 functionality? Or it requires two public IP addresses of exactly the same type? Lets assume we are talking about clients who support IPv6.
> 
Fro RFC5780 needs on 2 addresses from the same address family AFAIU. 
So it is not good one IPv6 and one IPv4 global ip for that.
> And finally a third question, given that it is now possible to use domain name in the `external-ip` field (as explained [here](https://github.com/coturn/coturn/issues/563#issuecomment-747190232)), could we maybe build on top of this, and map the first resolved A record to `external-ip` v4 and the first resolved AAAA record to `external-ip` v6? That way we can specify `external-ip=turn.example.com` and this would use both IPv4 and IPv6 if A and AAAA records are set for `turn.example.com` domain?
> 
make_ioa_addr function that is called first try to understand string as ipv4 then IPv6 and fallback only to dns lastly. The dns resolve first ipv4 and if fails only after it ipv6.

We can do to split external IP to IPv4 and IPv6, but according RFC recommendation you should not NAT IPv6. 
IMHO I still don't see strong use case for it. So I don't have strong opinion, but I tend to say we should implement it. 
Please describe why you want to NAT IPv6.

**Comment by maximbaz** [comment]
Many thanks for your detailed reply @misi! This is very interesting!

While you are correct that in most cases IPv6 NAT is not needed, a case for when it's still useful in Docker is best described in [`docker-ipv6nat` README](https://github.com/robbertkl/docker-ipv6nat#why-would-i-need-this) - maybe not all of those reasons apply to `coturn` specifically, but for example I run `coturn` on the same single server where I also host other apps for which those reasons do apply, and so I already have NAT configured for Docker.

> AFAIK docker assign IP when it is starting a docker container. 

This is correct.

> So e.g. in ENTYPOINT shell script you have in environment var the local ip by default

This would have been nice, but unfortunately it doesn't seem to happen at least in my tests 😞 I guess I could try to do what coturn does, in the ENTRYPOINT enumerate all existing IP addresses, try to detect the relevant ones, but this will become more complicated than it probably should be...

> you can also could get as extra ENV var the global IP too, and can replace in config before starting coturn.

This part wouldn't be a problem at all, I agree 👍 The biggest challenge is getting the correct internal IP address from inside the container, which changes on every restart...

> Fro RFC5780 needs on 2 addresses from the same address family AFAIU.
> So it is not good one IPv6 and one IPv4 global ip for that.

Good to know! While I was testing this, I specified mapping for one IPv4 and one IPv6 address using the extended syntax, and the message "WARNING: I cannot support STUN CHANGE_REQUEST functionality because only one IP address is provided" disappeared - it's a minor thing, but seems like we should still print this message in such case.

In summary, I understand that using NAT for IPv6 is not a very common thing to do, but in my setup it is beneficial just because it is not very good (yet?) to work with IPv6 in Docker. If you think we should close this ticket, I will implement this using a script that detects local IP addresses on startup 👍

**Comment by maximbaz** [comment]
It turned out to be a lot easier to detect the private IP addresses than I expected, so I implemented the suggestion you gave me, by creating a custom entrypoint for the docker. I leave it up to you to decide if this should be implemented in coturn itself, or you think supporting NAT IPv6 is out of scope for the project 🙂 

```sh
#!/bin/bash

apt install jq -y

ip4="$(ip --json addr | jq -r '.[] | .addr_info[] | select (.scope == "global" and .family == "inet") | .local')"
ip6="$(ip --json addr | jq -r '.[] | .addr_info[] | select (.scope == "global" and .family == "inet6") | .local')"

echo >&2 "Mapping IP addresses:"
echo >&2 "private IPv4 '$ip4' => public IPv4 '$PUBLIC_IP4'"
echo >&2 "private IPv6 '$ip6' => public IPv6 '$PUBLIC_IP6'"
echo >&2
echo >&2 "==== Starting coturn ===="
echo >&2

sed "s/PRIVATE_IP4/$ip4/g; s/PRIVATE_IP6/$ip6/g; s/PUBLIC_IP4/$PUBLIC_IP4/g; s/PUBLIC_IP6/$PUBLIC_IP6/g" /etc/turnserver.conf.tpl > /etc/turnserver.conf

/usr/local/bin/turnserver
```

**Comment by misi** [comment]
Thanks! Sorry I mixed k8s and docker. 
So you don't get ip by default on docker, but you can expose  status.podIP in k8s
https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/

Don't get me wrong. If you see worth to implement IPv6 NAT and you can make the PR, I am happy to include it..
Unfortunately I have limited time and I don't think this use case is very common and there are more urgent things to fix. 

Thanks for your detailed explanation and for the feedback about your docker solution!

**Comment by maximbaz** [comment]
Understood, and agree! Let's close the issue not to take up the space, if I get to implementing the PR I will just send it directly 🙂👍
